File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/sorting/ShellSort.java
package com.sortingsearching.sorting;

import java.util.Arrays;

public final class ShellSort {
    private ShellSort() {
    }

    public static int[] sort(int[] input) {
        int[] values = Arrays.copyOf(input, input.length);
        for (int gap = values.length / 2; gap > 0; gap /= 2) {
            for (int i = gap; i < values.length; i++) {
                int current = values[i];
                int j = i;
                while (j >= gap && values[j - gap] > current) {
                    values[j] = values[j - gap];
                    j -= gap;
                }
                values[j] = current;
            }
        }
        return values;
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 7, 99, 1};
        System.out.println("ShellSort");
        System.out.println("input : " + Arrays.toString(sample));
        System.out.println("output: " + Arrays.toString(sort(sample)));
    }
}

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/sorting/shell_sort.py
def shell_sort(values: list[int]) -> list[int]:
    result = values[:]
    gap = len(result) // 2

    while gap > 0:
        for i in range(gap, len(result)):
            current = result[i]
            j = i
            while j >= gap and result[j - gap] > current:
                result[j] = result[j - gap]
                j -= gap
            result[j] = current
        gap //= 2

    return result


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 7, 99, 1]
    print("ShellSort")
    print(f"input : {sample}")
    print(f"output: {shell_sort(sample)}")
